---
title: "Create_haplotypes"
author: "Shawn"
date: '2024-05-04'
output: html_document
---

Set working directory

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "C:/Users/shawn0601/Downloads/STP_malaria_data_scripts/Drug_resistance/Data")
```

Import libraries

```{r libs, echo=FALSE, warning = FALSE, message= FALSE}
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyr)
library(stringr)
```

Evaluate average mutation frequency in phenotype data

```{r average mutation frequency}

resm_snp <- read.csv("resm_data_clean_v2_v0.1.8.csv")

resm_phenotype <- resm_snp %>%
  group_by(dataset, SampleID, marker, Codon, CodonRefAlt, AA, AARefAlt) %>%
  mutate(Sum_reads = sum(Reads)) %>% # some sample has multiple REFERENCE alleles
  distinct(dataset, SampleID, marker, Codon, CodonRefAlt, AA, AARefAlt, Sum_reads, .keep_all = T) %>% # remove these samples
  group_by(dataset, SampleID, marker) %>%
  mutate(MjMn_prop = Sum_reads/sum(Sum_reads)) %>%
  mutate_at(vars(MjMn_prop), ~ replace(., is.nan(.), 1)) %>%
  mutate(Pheno = case_when(
           MjMn_prop == 1 & AARefAlt == "REF" ~ "Wild-type", 
           MjMn_prop == 1 & AARefAlt == "ALT" ~ "Mutant",
           MjMn_prop != 1 ~ "Mixed", T ~ NA),
         Scoring = case_when(
           Pheno == "Wild-type" ~ 0, 
           Pheno == "Mutant" ~ 1, 
           Pheno == "Mixed" & AARefAlt == "REF" ~ 0,
           Pheno == "Mixed" & AARefAlt == "ALT" ~ MjMn_prop,  T ~ NA)) %>%
  group_by(dataset, SampleID, marker, AARefAlt) %>%
  mutate(Total_scoring = sum(Scoring))

ave_mut_freq <- resm_phenotype %>%
  group_by(marker, SampleID) %>%
  filter(Scoring == max(Scoring)) %>%
  group_by(marker) %>%
  summarise(mut_freq =  mean(Scoring),
            uqt = quantile(Scoring, 0.975),
            lqt = quantile(Scoring, 0.025),
            n = n(),
            sd = sd(Scoring))

```


Transform mutant phenotype to haplotypes data (only major haplotypes)

```{r transform to haplotypes}

#### dhfr-dhps haplotypes ####
dd_haplo <- resm_phenotype %>%
  group_by(marker, SampleID) %>%
  filter(MjMn_prop == max(MjMn_prop)) %>%
  filter(Gene == "dhps" | Gene == "dhfr") %>%
  filter(marker %in% (ave_mut_freq$marker[ave_mut_freq$mut_freq != 0])) %>%
  mutate(type = ifelse(AARefAlt == "REF", "", "R")) %>%
  group_by(dataset, SampleID, Gene) %>%
  mutate(hap = paste0(CodonID, collapse = "/"),
         AA = paste0(AA, collapse = ""),
         lv = paste0(type, collapse = "")) %>%
  distinct(dataset, SampleID, hap, AA, lv) %>%
  group_by(dataset, SampleID) %>%
  mutate(dd_hap = paste0(hap, collapse = "-"),
         dd_AA = paste0(AA, collapse = "-"),
         Res_lv = paste0(lv, collapse = "-"),
         hap_r_num = str_count(Res_lv, "R")) %>%
  mutate(Res_lv = ifelse(endsWith(Res_lv, "R") == F, str_c(Res_lv, "WT"), Res_lv)) %>%
  distinct(dataset, SampleID, dd_hap, dd_AA, Res_lv, hap_r_num)

#### mdr1 haplotypes ####
mdr1_haplo <- resm_phenotype %>%
  group_by(marker, SampleID) %>%
  filter(MjMn_prop == max(MjMn_prop)) %>%
  filter(Gene == "mdr1") %>%
  filter(marker %in% (ave_mut_freq$marker[ave_mut_freq$mut_freq != 0])) %>%
  mutate(type = ifelse(AARefAlt == "REF", "", "R")) %>%
  group_by(dataset, SampleID, Gene) %>%
  mutate(mdr1_hap = paste0(CodonID, collapse = "/"),
         mdr1_AA = paste0(AA, collapse = ""),
         Res_lv = paste0(type, collapse = ""),
         hap_r_num = str_count(Res_lv, "R")) %>%
  mutate(Res_lv = ifelse(endsWith(Res_lv, "R") == F, str_c(Res_lv, "WT"), Res_lv)) %>%
  distinct(dataset, SampleID, mdr1_hap, mdr1_AA, Res_lv, hap_r_num)

#### crt haplotypes ####
crt_haplo <- resm_phenotype %>%
  group_by(marker, SampleID) %>%
  filter(MjMn_prop == max(MjMn_prop)) %>%
  filter(Gene == "crt") %>%
  filter(marker %in% (ave_mut_freq$marker[ave_mut_freq$mut_freq != 0])) %>%
  mutate(type = ifelse(AARefAlt == "REF", "", "R")) %>%
  group_by(dataset, SampleID, Gene) %>%
  mutate(crt_hap = paste0(CodonID, collapse = "/"),
         crt_AA = paste0(AA, collapse = ""),
         Res_lv = paste0(type, collapse = ""),
         hap_r_num = str_count(Res_lv, "R")) %>%
  mutate(Res_lv = ifelse(endsWith(Res_lv, "R") == F, str_c(Res_lv, "WT"), Res_lv)) %>%
  distinct(dataset, SampleID, crt_hap, crt_AA, Res_lv, hap_r_num)

write.csv(dd_haplo, "dhfr_dhps_haplotypes.csv", row.names = F)
write.csv(mdr1_haplo, "mdr1_haplotypes.csv", row.names = F)
write.csv(crt_haplo, "crt_haplotypes.csv", row.names = F)

```

