---
title: "Resistance_QC"
author: "Shawn"
date: '2024-05-03'
output: html_document
---

Set working directory

```{r setup, include=FALSE}
knitr::opts_knit$set(root.dir = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script")
```

Import libraries

```{r libs, echo=FALSE, warning = FALSE, message= FALSE}
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyr)
library(stringr)
```

Cleaning resistance marker data

```{r clean resm data}

resm_dt <- read.csv("resm_data_concatenated.csv")


#### Clean 1: remove controls, and cases outside the 2010-2016 study period OR non-D0 OR QC fail sequencing run ####
pc <- resm_dt$SampleID[grepl("PC", resm_dt$SampleID)]
nc <- resm_dt$SampleID[grepl("NC", resm_dt$SampleID)]
controls <- c(pc, nc)
dt_05 <- resm_dt$SampleID[grepl("2005-", resm_dt$SampleID)]
dt_07 <- resm_dt$SampleID[grepl("2007-", resm_dt$SampleID)]
dt_not_D0 <- resm_dt$SampleID[grepl("D3", resm_dt$SampleID)]
qc_fail_seq <- c("M07", "M09", "M10", "M13", "M15")

resm_clean1 <- resm_dt %>%
  filter(!SampleID %in% controls & !SampleID %in% dt_05 & !SampleID %in% dt_07 & !SampleID %in% dt_not_D0) %>%
  mutate(dataset = gsub("_resmarker_table", "", dataset)) %>%
  filter(! dataset %in% qc_fail_seq) %>%
  mutate(SampleID = case_when(
    dataset == "NS1" | dataset == "NS2" ~ str_sub(str_remove(SampleID, "[:punct:][S][:digit:]+$"),36,-1),
    TRUE ~ gsub(".*_","", SampleID)))


#### Clean 2: remove QC check samples by sum of reads less than 10 in half of SNP ####
resm_clean2 <- resm_clean1 %>%
  mutate(marker = str_c(Gene, "_", CodonID)) %>%
  mutate_at(c("Reads"), as.numeric) %>%
  group_by(dataset, SampleID, marker) %>%
  mutate(Sum_reads = sum(Reads)) %>%
  filter(marker != "crt_326") %>% # remove SNP only included in pool 2
  ungroup() %>%
  distinct(dataset, SampleID, marker, Sum_reads) %>%
  pivot_wider(names_from = marker, values_from = Sum_reads) %>%
  mutate(count_NA = rowSums(is.na(.[,3:70])),
         count_less_10 = rowSums(.[,3:70] < 10, na.rm = T),
         count_btw_10_100 = rowSums(.[,3:70] < 100, na.rm = T),
         count_btw_0_10 = count_NA + count_less_10) %>%
  filter(!count_btw_0_10 > 68/2)


#### Clean 3: remove repeated sequencing samples by comparing the total reads and reads condition ####
n_occur <- data.frame(table(distinct(resm_clean2, dataset, SampleID)[, c("SampleID")]))
dup_list <- n_occur[n_occur$Freq > 1,]

dup_id_select <- resm_clean2 %>%
  filter(SampleID %in% dup_list$SampleID) %>%
  group_by(dataset, SampleID) %>%
  cbind(Total_reads = rowSums(.[,3:70], na.rm = T)) %>%
  as.data.frame(.[,1:75]) %>%
  distinct(dataset, SampleID, Total_reads, count_btw_0_10) %>%
  group_by(SampleID) %>%
  filter(count_btw_0_10 == min(count_btw_0_10)) %>%
  filter(Total_reads == max(Total_reads)) %>%
  mutate(id = str_c(dataset, "_", SampleID))

dup_id_not_select <- resm_clean2 %>%
  filter(SampleID %in% dup_list$SampleID) %>%
  summarise(id = str_c(dataset, "_", SampleID)) %>%
  filter(!id %in% dup_id_select$id)

resm_clean3 <- resm_clean2 %>%
  mutate(id = str_c(dataset, "_", SampleID)) %>%
  filter(!id %in% dup_id_not_select$id) %>%
  select(-c(71:75)) %>%
  replace(is.na(.), 0) %>%
  pivot_longer(cols = c(3:70), names_to = "marker", values_to = "Sum_reads") %>%
  separate(marker, c("Gene", "CodonID"), "_") %>%
  mutate_at("CodonID", as.numeric) %>%
  merge(resm_clean1, by = c("dataset", "SampleID", "Gene", "CodonID"), all.x = T) %>%
  left_join(distinct(resm_clean1, Gene, CodonID, GeneID, RefCodon, CodonStart, RefAA), by = c("Gene", "CodonID"), suffix = c("", ".2")) %>%
  mutate(GeneID = ifelse(is.na(GeneID), GeneID.2, GeneID),
         RefCodon = ifelse(is.na(RefCodon), RefCodon.2, RefCodon),
         CodonStart = ifelse(is.na(CodonStart), CodonStart.2, CodonStart),
         RefAA = ifelse(is.na(RefAA), RefAA.2, RefAA),
         Codon = ifelse(Sum_reads == 0, RefCodon, Codon),
         CodonRefAlt = ifelse(Sum_reads == 0, "REF", CodonRefAlt),
         AA = ifelse(Sum_reads == 0, RefAA, AA),
         AARefAlt = ifelse(Sum_reads == 0, "REF", AARefAlt),
         Reads =  ifelse(Sum_reads == 0, 0, Reads)) %>%
  select(-c(GeneID.2, RefCodon.2, CodonStart.2, RefAA.2))


#### Modifying: dhps 437 (reference G is resistant type, so re-coded as A437G) & k13 454 site (codon and amino acid error) ####
resm_fix <- resm_clean3 %>%
  mutate(marker = str_c(Gene, "_", CodonID)) %>%
  mutate(
    RefCodon = 
      case_when(marker == "dhps_437" ~ "GCT", marker == "k13_454" ~ "GTA", T ~ RefCodon),
    Codon = 
      case_when(marker == "k13_454" ~ "GTA", T ~ Codon),
    CodonStart = 
      case_when(marker == "k13_454" ~ 55, T ~ CodonStart),
    CodonRefAlt = 
      case_when(marker == "dhps_437" & Codon == "GCT" ~ "REF", marker == "dhps_437" & Codon == "GGT" ~ "ALT", T ~ CodonRefAlt),
    RefAA = 
      case_when(marker == "dhps_437" ~ "A", marker == "k13_454" ~ "V", T ~ RefAA),
    AA = 
      case_when(marker == "k13_454" ~ "V", T ~ AA),
    AARefAlt = 
      case_when(marker == "dhps_437" & AA == "A" ~ "REF", marker == "dhps_437" & AA == "G" ~ "ALT", T ~ AARefAlt)) %>%
  relocate(marker, .before = Gene)


write.csv(resm_fix, "resm_data_clean_v2_v0.1.8.csv", row.names = F)

```
