---
title: "13.NeEstimator_Ne_preprocess"
author: "Shawn"
date: '2024-03-14'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script")
```

Import libraries.

```{r libs}
library(scales)
library(data.table)
library(plotrix)
library(scales)
library(gtools)
library(timetk)
library(pheatmap)
library(hierfstat)
library(dplyr)
library(magrittr)
library(ggplot2)
library(tidyr)
library(waldo)
library(stringr)
library(tools)
library(purrr)
library(parallel)
library(data.table)
library(vctrs)
library(plotly)
```

Output fstat data for NeEstimator.

```{r diversity trial 5 times with all have-changed snp, include=TRUE}

ALL_allele_align <- read.csv("ALL_allele_aligned_with_reference_seq.csv")
div_allele_align <- ALL_allele_align %>% filter(Category == "Diversity")

# Set seed to create multiple RANDOM dataset
div_SEED_vector <- c(1661, 1662, 1663, 1664, 1665)

for (my_seed in div_SEED_vector[1:5]){
  # set seed
  set.seed(my_seed)
  print(paste0("Seed: ", my_seed))
  
  # Filter major locus and RANDOM loci with same proportion
  div_allele_align_filtered <- div_allele_align %>%
    group_by(sampleID, locus_new) %>%
    filter(norm.reads.locus == max(norm.reads.locus)) %>% # only the dominant allele is selected
    slice_sample(n=1) %>% # random select a row with same locus proportion
    ungroup() %>%
    group_by(locus_new) %>%
    mutate(asv_align_remove_mismatch = str_pad(asv_align_remove_mismatch, width = max(asv_final_length), side = 'right', pad = '-'),
           asv_length_in_loci = nchar(asv_align_remove_mismatch)) %>% # fill-in the empty cells without sequences
    pivot_wider(id_cols = sampleID, names_from = locus_new, values_from = asv_align_remove_mismatch) %>%
    pivot_longer(cols = !sampleID, names_to = "locus_new", values_to = "asv_align_remove_mismatch") %>%
    group_by(locus_new) %>%
    mutate_at("asv_align_remove_mismatch", ~replace_na(., paste(rep("-", max(nchar(asv_align_remove_mismatch), na.rm = T)), collapse = ""))) %>% 
    merge(distinct(div_allele_w_locus_pools[,c("locus_new", "Chromosome", "Amplicon_start_new")]), by = "locus_new")
  print("Step 1: filter DONE!")
  
  # Split aligned dataframe by chromosome to reduce the runtime
  chr_list_div_allele_align <- split(div_allele_align_filtered, div_allele_align_filtered$Chromosome)
  
  # Split the sequence site by site and ONLY CHANGED SITE
  div_split_list <- list()
  for (i in 1:length(chr_list_div_allele_align)) {
    x <- chr_list_div_allele_align[[i]]
    splited <- x %>%
      group_by(sampleID, locus_new, asv_align_remove_mismatch, Amplicon_start_new) %>%
      mutate(seq = paste(strsplit(as.character(asv_align_remove_mismatch), "")[[1]], collapse = ", "),
             pos = paste(seq(from = Amplicon_start_new+1, 
                             to = Amplicon_start_new+nchar(asv_align_remove_mismatch), by=1), 
                         collapse=", ")) %>%
      separate_longer_delim(c(seq, pos), delim = ", ") %>%
      mutate(site = str_c(Chromosome, "_", pos))
    changed_site <- splited %>%
      ungroup() %>%
      filter(seq != "-") %>%
      group_by(locus_new, site, seq) %>%
      summarise(n = n(), .groups = "drop") %>%
      group_by(locus_new, site) %>%
      mutate(pct = n/sum(n)) %>%
      filter(pct != 1)
    missing_site <- splited %>%
      ungroup() %>%
      group_by(locus_new, site, seq) %>%
      summarise(n = n(), .groups = "drop") %>%
      group_by(locus_new, site) %>%
      mutate(pct = n/sum(n)) %>%
      filter(seq == "-" & pct >= 0.5)
    splited_only_changed_site <- splited %>%
      ungroup() %>%
      select(sampleID, Chromosome, site, Amplicon_start_new, seq) %>%
      filter((site %in% changed_site$site) & (!site %in% missing_site$site)) %>%
      group_by(Chromosome, site, sampleID) %>%
      filter(Amplicon_start_new == min(Amplicon_start_new))
    div_split_list[[i]] <- splited_only_changed_site
  }
  print("Step 2: only changed site DONE!")
  
  #Transform to genotypes coding (A=1, T=2, G=3, C=4)
  div_geno_list <- list()
  for (i in 1:length(div_split_list)){
    x <- div_split_list[[i]]
    geno_table <- x %>%
      group_by(sampleID, Chromosome, site, Amplicon_start_new, seq) %>%
      mutate(coding = case_when(seq == "A" ~ 11, seq == "T" ~ 22,
                                seq == "G" ~ 33, seq == "C" ~ 44, T ~ 00)) %>%
      select(sampleID, site, coding) %>%
      pivot_wider(id_cols = sampleID, names_from = site, values_from = coding) %>%
      arrange(sampleID)
    div_geno_list[[i]] <- geno_table
  }
  print("Step 3: genotypes coding DONE!")

  div_allele_geno <- div_geno_list[[1]]
  for (i in 2:length(div_geno_list)) {
    x <- div_geno_list[[i]]
    div_allele_geno <- left_join(div_allele_geno, x, by = "sampleID")
  }

  # Adding time columns
  div_allele_genotype_for_ne <- merge(div_allele_geno, distinct(meta_data[, c("DNA_tube_ID", "time_gp")]), by.x = c("sampleID"), by.y = c("DNA_tube_ID"), all.x = T) %>% relocate("time_gp", .after = "sampleID")

  div_genotype_10_11 <- subset(div_allele_genotype_for_ne, time_gp %in% c(2010, 2011))%>%
    mutate(pop = ifelse(time_gp == min(time_gp), 1, 2)) %>%
    relocate("pop", .after = "time_gp") %>%
    arrange(pop) %>% select(-c(1,2)) %>% na.replace(00)

  div_genotype_11_12 <- subset(div_allele_genotype_for_ne, time_gp %in% c(2011, 2012)) %>%
    mutate(pop = ifelse(time_gp == min(time_gp), 1, 2)) %>%
    relocate("pop", .after = "time_gp") %>%
    arrange(pop) %>% select(-c(1,2)) %>% na.replace(00)

  div_genotype_12_13 <- subset(div_allele_genotype_for_ne, time_gp %in% c(2012, 2013)) %>%
    mutate(pop = ifelse(time_gp == min(time_gp), 1, 2)) %>%
    relocate("pop", .after = "time_gp") %>%
    arrange(pop) %>% select(-c(1,2)) %>% na.replace(00)

  div_genotype_13_14 <- subset(div_allele_genotype_for_ne, time_gp %in% c(2013, 2014)) %>%
    mutate(pop = ifelse(time_gp == min(time_gp), 1, 2)) %>%
    relocate("pop", .after = "time_gp") %>%
    arrange(pop) %>% select(-c(1,2)) %>% na.replace(00)

  div_genotype_14_15 <- subset(div_allele_genotype_for_ne, time_gp %in% c(2014, 2015)) %>%
    mutate(pop = ifelse(time_gp == min(time_gp), 1, 2)) %>%
    relocate("pop", .after = "time_gp") %>%
    arrange(pop) %>% select(-c(1,2)) %>% na.replace(00)

  div_genotype_15_16 <- subset(div_allele_genotype_for_ne, time_gp %in% c(2015, 2016)) %>%
    mutate(pop = ifelse(time_gp == min(time_gp), 1, 2)) %>%
    relocate("pop", .after = "time_gp") %>%
    arrange(pop) %>% select(-c(1,2)) %>% na.replace(00)
  print("Step 4: adding time columns DONE!")

  write.fstat(div_genotype_10_11, paste0("div_genotype_10_11_", my_seed, ".dat"))
  write.fstat(div_genotype_11_12, paste0("div_genotype_11_12_", my_seed, ".dat"))
  write.fstat(div_genotype_12_13, paste0("div_genotype_12_13_", my_seed, ".dat"))
  write.fstat(div_genotype_13_14, paste0("div_genotype_13_14_", my_seed, ".dat"))
  write.fstat(div_genotype_14_15, paste0("div_genotype_14_15_", my_seed, ".dat"))
  write.fstat(div_genotype_15_16, paste0("div_genotype_15_16_", my_seed, ".dat"))
  print("Step 5: write fstat file DONE!")

  # Create frequency file (all)
  div_snp_freq <- list()
  for (i in 1:length(div_split_list)) {
    freq_selected_site <- div_split_list[[i]] %>%
      ungroup() %>%
      filter(seq != "-") %>%
      group_by(Chromosome, site, seq) %>%
      summarise(n = n(), .groups = "drop") %>%
      group_by(Chromosome, site) %>%
      mutate(pct = n/sum(n))
    div_snp_freq[[i]] <- freq_selected_site
  }
  div_snp_freq_dt = do.call(rbind, div_snp_freq)
  write.csv(div_snp_freq_dt, paste0("div_snp_frequency_", my_seed, ".csv"), row.names = F)
  
  hist_div_snp_freq <- div_snp_freq_dt %>%
    ggplot(aes(x = pct)) +
    geom_histogram(binwidth = 0.01) +
    scale_y_continuous(trans = "mylog10") +
    ggtitle(paste0("diversity random site frequency: ", my_seed)) +
    xlab("allele frequency")
  ggsave(filename = paste0("div_random_site_freq_", my_seed, ".png"))
  print("Step 6: create frequency file DONE!")
  
}


```

Visualize ne estimation result

```{r visualize ne, include=TRUE}

#### 1661 ####
# Input
div_filedir_1661 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1661_TpxTp", pattern = ".txt$", ignore.case = T, full.names = T, recursive = T)
div_filename_1661 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1661_TpxTp", pattern = ".txt$", ignore.case = T, full.names = F, recursive = T)

div_ne_results_list_1661 <- data.frame(div_filename_1661, div_filedir_1661)
div_ne_results_list_1661$div_filename_1661 <- gsub("div_genotype_", "", gsub("_1661", "", gsub("_TpxTp.txt", "", div_ne_results_list_1661$div_filename_1661)))


div_ne_results_modif_list_1661 <- list()
for (i in 1:nrow(div_ne_results_list_1661)) {
  x <- readLines(div_ne_results_list_1661[i,2])
  y <- gsub("\\s+", " ", gsub("&", "", str_trim(unlist(strsplit(x,"\t"))[16:18]))) %>% as.data.frame()
  y[c("Sample_Pair_IDs", "Start_Gen", "End_Gen", "Method", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper")] <- str_split_fixed(y$., ' ', 13)
  y <- select(y, -1)
  div_ne_results_modif_list_1661[[i]] <- y
  names(div_ne_results_modif_list_1661)[i] <- div_ne_results_list_1661[i,1]
}
div_ne_results_modif_1661 <- rbindlist(div_ne_results_modif_list_1661, idcol=TRUE) %>%
  mutate(Year = case_when(
    str_detect(.id, "10_11")==T ~ 2011, str_detect(.id, "11_12")==T ~ 2012,
    str_detect(.id, "12_13")==T ~ 2013, str_detect(.id, "13_14")==T ~ 2014,
    str_detect(.id, "14_15")==T ~ 2015, str_detect(.id, "15_16")==T ~ 2016, 
    T ~ NA
  )) %>%
  mutate_at(c("Start_Gen", "End_Gen", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper"),as.numeric)

# Visualize
p_div_ne_change_over_generation_1661 <- div_ne_results_modif_1661 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = End_Gen, y = Ne, fill = factor(Year), color = factor(Year))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1661") +
  xlab("Generation") +
  ylab("Effective population size") +
  labs(color = "Year") +
  coord_cartesian(ylim=c(0, 110)) +
  scale_x_continuous(breaks= pretty_breaks()) +
  theme_classic() +
  theme(text = element_text(size = 20))


p_div_ne_change_over_year_1661 <- div_ne_results_modif_1661 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = factor(Year,levels = unique(Year)), y = Ne, fill = factor(End_Gen), color = factor(End_Gen), group = factor(End_Gen))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1661") +
  xlab("Year") +
  ylab("Effective population size") +
  labs(color = "Generation") +
  coord_cartesian(ylim=c(0, 110)) +
  theme_classic() +
  theme(text = element_text(size = 20))

#### 1662 ####
# Input
div_filedir_1662 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1662_TpxTp", pattern = ".txt$", ignore.case = T, full.names = T, recursive = T)
div_filename_1662 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1662_TpxTp", pattern = ".txt$", ignore.case = T, full.names = F, recursive = T)

div_ne_results_list_1662 <- data.frame(div_filename_1662, div_filedir_1662)
div_ne_results_list_1662$div_filename_1662 <- gsub("div_genotype_", "", gsub("_1662", "", gsub("_TpxTp.txt", "", div_ne_results_list_1662$div_filename_1662)))


div_ne_results_modif_list_1662 <- list()
for (i in 1:nrow(div_ne_results_list_1662)) {
  x <- readLines(div_ne_results_list_1662[i,2])
  y <- gsub("\\s+", " ", gsub("&", "", str_trim(unlist(strsplit(x,"\t"))[16:18]))) %>% as.data.frame()
  y[c("Sample_Pair_IDs", "Start_Gen", "End_Gen", "Method", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper")] <- str_split_fixed(y$., ' ', 13)
  y <- select(y, -1)
  div_ne_results_modif_list_1662[[i]] <- y
  names(div_ne_results_modif_list_1662)[i] <- div_ne_results_list_1662[i,1]
}
div_ne_results_modif_1662 <- rbindlist(div_ne_results_modif_list_1662, idcol=TRUE) %>%
  mutate(Year = case_when(
    str_detect(.id, "10_11")==T ~ 2011, str_detect(.id, "11_12")==T ~ 2012,
    str_detect(.id, "12_13")==T ~ 2013, str_detect(.id, "13_14")==T ~ 2014,
    str_detect(.id, "14_15")==T ~ 2015, str_detect(.id, "15_16")==T ~ 2016, 
    T ~ NA
  )) %>%
  mutate_at(c("Start_Gen", "End_Gen", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper"),as.numeric)

# Visualize
p_div_ne_change_over_generation_1662 <- div_ne_results_modif_1662 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = End_Gen, y = Ne, fill = factor(Year), color = factor(Year))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1662") +
  xlab("Generation") +
  ylab("Effective population size") +
  labs(color = "Year") +
  coord_cartesian(ylim=c(0, 160)) +
  scale_x_continuous(breaks= pretty_breaks()) +
  theme_classic() +
  theme(text = element_text(size = 20))


p_div_ne_change_over_year_1662 <- div_ne_results_modif_1662 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = factor(Year,levels = unique(Year)), y = Ne, fill = factor(End_Gen), color = factor(End_Gen), group = factor(End_Gen))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1662") +
  xlab("Year") +
  ylab("Effective population size") +
  labs(color = "Generation") +
  coord_cartesian(ylim=c(0, 160)) +
  theme_classic() +
  theme(text = element_text(size = 20))

#### 1663 ####
# Input
div_filedir_1663 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1663_TpxTp", pattern = ".txt$", ignore.case = T, full.names = T, recursive = T)
div_filename_1663 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1663_TpxTp", pattern = ".txt$", ignore.case = T, full.names = F, recursive = T)

div_ne_results_list_1663 <- data.frame(div_filename_1663, div_filedir_1663)
div_ne_results_list_1663$div_filename_1663 <- gsub("div_genotype_", "", gsub("_1663", "", gsub("_TpxTp.txt", "", div_ne_results_list_1663$div_filename_1663)))


div_ne_results_modif_list_1663 <- list()
for (i in 1:nrow(div_ne_results_list_1663)) {
  x <- readLines(div_ne_results_list_1663[i,2])
  y <- gsub("\\s+", " ", gsub("&", "", str_trim(unlist(strsplit(x,"\t"))[16:18]))) %>% as.data.frame()
  y[c("Sample_Pair_IDs", "Start_Gen", "End_Gen", "Method", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper")] <- str_split_fixed(y$., ' ', 13)
  y <- select(y, -1)
  div_ne_results_modif_list_1663[[i]] <- y
  names(div_ne_results_modif_list_1663)[i] <- div_ne_results_list_1663[i,1]
}
div_ne_results_modif_1663 <- rbindlist(div_ne_results_modif_list_1663, idcol=TRUE) %>%
  mutate(Year = case_when(
    str_detect(.id, "10_11")==T ~ 2011, str_detect(.id, "11_12")==T ~ 2012,
    str_detect(.id, "12_13")==T ~ 2013, str_detect(.id, "13_14")==T ~ 2014,
    str_detect(.id, "14_15")==T ~ 2015, str_detect(.id, "15_16")==T ~ 2016, 
    T ~ NA
  )) %>%
  mutate_at(c("Start_Gen", "End_Gen", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper"),as.numeric)

# Visualize
p_div_ne_change_over_generation_1663 <- div_ne_results_modif_1663 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = End_Gen, y = Ne, fill = factor(Year), color = factor(Year))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1663") +
  xlab("Generation") +
  ylab("Effective population size") +
  labs(color = "Year") +
  coord_cartesian(ylim=c(0, 160)) +
  scale_x_continuous(breaks= pretty_breaks()) +
  theme_classic() +
  theme(text = element_text(size = 20))


p_div_ne_change_over_year_1663 <- div_ne_results_modif_1663 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = factor(Year,levels = unique(Year)), y = Ne, fill = factor(End_Gen), color = factor(End_Gen), group = factor(End_Gen))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1663") +
  xlab("Year") +
  ylab("Effective population size") +
  labs(color = "Generation") +
  coord_cartesian(ylim=c(0, 160)) +
  theme_classic() +
  theme(text = element_text(size = 20))

#### 1664 ####
# Input
div_filedir_1664 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1664_TpxTp", pattern = ".txt$", ignore.case = T, full.names = T, recursive = T)
div_filename_1664 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1664_TpxTp", pattern = ".txt$", ignore.case = T, full.names = F, recursive = T)

div_ne_results_list_1664 <- data.frame(div_filename_1664, div_filedir_1664)
div_ne_results_list_1664$div_filename_1664 <- gsub("div_genotype_", "", gsub("_1664", "", gsub("_TpxTp.txt", "", div_ne_results_list_1664$div_filename_1664)))


div_ne_results_modif_list_1664 <- list()
for (i in 1:nrow(div_ne_results_list_1664)) {
  x <- readLines(div_ne_results_list_1664[i,2])
  y <- gsub("\\s+", " ", gsub("&", "", str_trim(unlist(strsplit(x,"\t"))[16:18]))) %>% as.data.frame()
  y[c("Sample_Pair_IDs", "Start_Gen", "End_Gen", "Method", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper")] <- str_split_fixed(y$., ' ', 13)
  y <- select(y, -1)
  div_ne_results_modif_list_1664[[i]] <- y
  names(div_ne_results_modif_list_1664)[i] <- div_ne_results_list_1664[i,1]
}
div_ne_results_modif_1664 <- rbindlist(div_ne_results_modif_list_1664, idcol=TRUE) %>%
  mutate(Year = case_when(
    str_detect(.id, "10_11")==T ~ 2011, str_detect(.id, "11_12")==T ~ 2012,
    str_detect(.id, "12_13")==T ~ 2013, str_detect(.id, "13_14")==T ~ 2014,
    str_detect(.id, "14_15")==T ~ 2015, str_detect(.id, "15_16")==T ~ 2016, 
    T ~ NA
  )) %>%
  mutate_at(c("Start_Gen", "End_Gen", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper"),as.numeric)

# Visualize
p_div_ne_change_over_generation_1664 <- div_ne_results_modif_1664 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = End_Gen, y = Ne, fill = factor(Year), color = factor(Year))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1664") +
  xlab("Generation") +
  ylab("Effective population size") +
  labs(color = "Year") +
  coord_cartesian(ylim=c(0, 160)) +
  scale_x_continuous(breaks= pretty_breaks()) +
  theme_classic() +
  theme(text = element_text(size = 20))


p_div_ne_change_over_year_1664 <- div_ne_results_modif_1664 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = factor(Year,levels = unique(Year)), y = Ne, fill = factor(End_Gen), color = factor(End_Gen), group = factor(End_Gen))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1664") +
  xlab("Year") +
  ylab("Effective population size") +
  labs(color = "Generation") +
  coord_cartesian(ylim=c(0, 160)) +
  theme_classic() +
  theme(text = element_text(size = 20))

#### 1665 ####
# Input
div_filedir_1665 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1665_TpxTp", pattern = ".txt$", ignore.case = T, full.names = T, recursive = T)
div_filename_1665 <- list.files(path = "F:/Malaria_AmpSeq analysis/analysis/STP_data_script/no_N_1665_TpxTp", pattern = ".txt$", ignore.case = T, full.names = F, recursive = T)

div_ne_results_list_1665 <- data.frame(div_filename_1665, div_filedir_1665)
div_ne_results_list_1665$div_filename_1665 <- gsub("div_genotype_", "", gsub("_1665", "", gsub("_TpxTp.txt", "", div_ne_results_list_1665$div_filename_1665)))


div_ne_results_modif_list_1665 <- list()
for (i in 1:nrow(div_ne_results_list_1665)) {
  x <- readLines(div_ne_results_list_1665[i,2])
  y <- gsub("\\s+", " ", gsub("&", "", str_trim(unlist(strsplit(x,"\t"))[16:18]))) %>% as.data.frame()
  y[c("Sample_Pair_IDs", "Start_Gen", "End_Gen", "Method", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper")] <- str_split_fixed(y$., ' ', 13)
  y <- select(y, -1)
  div_ne_results_modif_list_1665[[i]] <- y
  names(div_ne_results_modif_list_1665)[i] <- div_ne_results_list_1665[i,1]
}
div_ne_results_modif_1665 <- rbindlist(div_ne_results_modif_list_1665, idcol=TRUE) %>%
  mutate(Year = case_when(
    str_detect(.id, "10_11")==T ~ 2011, str_detect(.id, "11_12")==T ~ 2012,
    str_detect(.id, "12_13")==T ~ 2013, str_detect(.id, "13_14")==T ~ 2014,
    str_detect(.id, "14_15")==T ~ 2015, str_detect(.id, "15_16")==T ~ 2016, 
    T ~ NA
  )) %>%
  mutate_at(c("Start_Gen", "End_Gen", "Harmonic_Mean_Size", "No_of_Indep_Alleles", "F", "F'", "Ne", "CIs_Parametric_lower", "CIs_Parametric_upper", "CIs_JackKnife_lower", "CIs_JackKnife_upper"),as.numeric)

# Visualize
p_div_ne_change_over_generation_1665 <- div_ne_results_modif_1665 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = End_Gen, y = Ne, fill = factor(Year), color = factor(Year))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1665") +
  xlab("Generation") +
  ylab("Effective population size") +
  labs(color = "Year") +
  coord_cartesian(ylim=c(0, 160)) +
  scale_x_continuous(breaks= pretty_breaks()) +
  theme_classic() +
  theme(text = element_text(size = 20))


p_div_ne_change_over_year_1665 <- div_ne_results_modif_1665 %>%
  filter(Method == "JR") %>%
  ggplot(aes(x = factor(Year,levels = unique(Year)), y = Ne, fill = factor(End_Gen), color = factor(End_Gen), group = factor(End_Gen))) +
  geom_line(linetype = "solid", position=position_dodge(width=0.5), linewidth=0.8) +
  geom_point(aes(shape = factor(Method)), position=position_dodge(width=0.5), show.legend = FALSE) +
  geom_errorbar(aes(ymin = CIs_Parametric_lower, ymax = CIs_Parametric_upper), width = 0.2, linetype = "dotted", position=position_dodge(width=0.5), show.legend = FALSE) +
  ggtitle("No. 1665") +
  xlab("Year") +
  ylab("Effective population size") +
  labs(color = "Generation") +
  coord_cartesian(ylim=c(0, 160)) +
  theme_classic() +
  theme(text = element_text(size = 20))


```
